### PR TITLE
Cantera version 2.6

### PR DESCRIPTION
**Pull request by ChrisBNEU**
### Motivation or Problem
addresses issue #2270. 
Cantera is deprecating the ".cti" output file type that RMG currently uses. for nicer features like coverage dependant thermo and kinetics, we need to upgrade cantera to v2.5 and switch to the yaml output file type. 

### Description of Changes
Changed the input parser from ck2cti to ck2yaml. there were some syntax changes for the parser from v2.3 to 2.5. 

### Testing
update all of the unit tests that involve writing or reading a cantera output file
ensure that a ct.Solution and ct.interface object load correctly externally. 

### Reviewer Tips
Any suggestions for what should be tested would be appreciated, I have not put any in.

### CODE REVIEW

**Comment by ChrisBNEU** [comment]
currently generated a chem_annotated.yaml and a chem.yaml. chem.yaml looks fine but the chem_annotated.yaml does not store the comments under "notes" for the reactions.

**Comment by ChrisBNEU** [comment]
This passes all tests and fixes the issue of not getting the coverage dependent parameters into the Cantera output file.

**Comment by ChrisBNEU** [comment]
One other thing to note is the test for generating an output file with the ck2yaml parser behaves a little differently. If we do not specify all of the elements used (e.g. we forget to add "C" for our carbon atom in the elements list) it will not throw an error. The error only appears when the output file is validated, which is not something cantera coded into the parser (you can do it with the command line ck2yaml parser however). I just changed the test case so it fails for another reason, since the point is not to exhaustively test every error possible when generating a yaml file. see this thread for more info: https://groups.google.com/g/cantera-users/c/eiTX-2O5zfQ

**Comment by ChrisBNEU** [comment]
upgrading to version 2.6 in environment.yml

**Comment by Laxzal** [comment]
Hey @ChrisBNEU, with this branch, are you experiencing an issue with installing PyCall in Julia (or add/installing new packages in Julia)

**Comment by Laxzal** [comment]
A fix for that issue is we need to make it use the `libcurl` packaged with Julia.

For my env, I had to do this
```
export LD_LIBRARY_PATH="/home/calvin/mambaforge/envs/rmg_env/share/julia/site/julia-1.6.7/lib/julia:$LD_LIBRARY_PATH"
```

**Comment by Laxzal** [comment]
> A fix for that issue is we need to make it use the `libcurl` packaged with Julia.
> 
> For my env, I had to do this
> 
> ```
> export LD_LIBRARY_PATH="/home/calvin/mambaforge/envs/rmg_env/share/julia/site/julia-1.6.7/lib/julia:$LD_LIBRARY_PATH"
> ```

@mjohnson541 - Return of the LD_LIBRARY_PATH issue

**Comment by JacksonBurns** [comment]
I am rebasing this branch to try and push this feature through now that the CI has been fixed.

**Comment by ChrisBNEU** [comment]
So there are a number of new issues with cantera 2.6 and beyond that have done my best to address:

Older versions of cantera used different `Reaction` objects for each reaction type (like `ArrheniusReaction`, `PlogReaction`, `InterfaceReaction`, etc.). There was a concerted effort to consolidate all reactions into the `cantera.Reaction` object, and differentiate with new `ReactionRate` objects (e.g. `LindemannRate`, `PlogRate`, `InterfaceArrheniusRate`). This was partially implemented in 2.6, but it is not fully implemented for `ThirdBodyReaction` types. These use the `ThirdBodyReaction` type, but with a `ThirdBodyRate`.

`FalloffReaction` is used for Lindemann and Troe reactions because it is what `cantera.Reaction.from_yaml` and `cantera.Solution(<yaml_file>)` creates in v2.6. In v3.0, `cantera.Reaction.from_yaml` will correctly create `cantera.Reaction` objects with `TroeRate` and `LindemannRate`. The reactionTest.py unit test is written such that it will only look at the rate, which will remain the same in v3.0

For versions <=3.0, we will not need to worry about third body reactions and falloff reactions. however, past version 3.0, we will need to recode so all reactions are type `cantera.Reaction`. Third body reactions will need a third_body object specified for the Reaction() object. For more info on reactions and rates in the newer release of cantera, refer to the [documentation](https://cantera.org/documentation/dev/sphinx/html/cython/kinetics.html#reactions).

All cti output file references are removed, since cti references are deprecated in v2.6, and will not work in later versions. additionally, cti files will not correctly write coverage dependencies, but yaml files will. 

I have not updated the surface reaction `to_cantera` methods. they did not exist before, and they require a fair bit of additional coding for unit conversion. I think that should live in a separate pr.

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
Thanks for the PR, implementation looks great! I have a couple minor comments on documentation and clarity.

One larger thing is that this does break compatibility with previous versions of RMG since we can no longer interface with the 'cti' format from Cantera. We should bundle this into a minor release of RMG-Py, like 3.3.0 or the like.

File: rmgpy/reaction.py [hunk 1]
```diff
@@ -245,6 +245,19 @@ def to_chemkin(self, species_list=None, kinetics=True):
         else:
             return rmgpy.chemkin.write_reaction_string(self)
 
+    def fix_reaction(self, ct_reactions, ct_products):
```
[the remark this answers is not in the document]
Comment (ChrisBNEU) [reply rmgpy/reaction.py:248]: I'm going to remove this, it was from Nora's yaml pr, but I decided not to address the surface reaction to_cantera methods in this pr. I just forgot to take it out.

File: rmgpy/kinetics/falloff.pyx [hunk 2]
```diff
@@ -228,9 +230,19 @@ cdef class Lindemann(PDepKineticsModel):
...
+
+    def to_cantera_kinetics(self): 
+        """
+        Converts the Lindemann object to a cantera LindemannRate object
+        """
+        import cantera as ct
+
+        high_rate = self.arrheniusHigh.to_cantera_kinetics(ArrObj=True)
+        low_rate = self.arrheniusLow.to_cantera_kinetics(ArrObj=True)
+        return ct.LindemannRate(low_rate, high_rate, [])
```
[the remark this answers is not in the document]
Comment (ChrisBNEU) [reply rmgpy/kinetics/falloff.pyx:244]: I'm putting this here so the question is answered, because the update I'm making does not pass a third argument. In Cantera, `LindemannRate`  can accept an additional falloff parameterization:
image

But if nothing is specified, it uses the same form rmg uses, shown below: 
image

File: rmgpy/kinetics/falloff.pyx [hunk 3]
```diff
@@ -389,15 +401,30 @@ cdef class Troe(PDepKineticsModel):
...
-            ct_reaction.falloff = ct.TroeFalloff(params=[A, T3, T1])
+            falloff = [A, T3, T1]
         else:
             T2 = self.T2.value_si
-            ct_reaction.falloff = ct.TroeFalloff(params=[A, T3, T1, T2])
+            falloff = [A, T3, T1, T2]
+
+        ct_reaction.rate = self.to_cantera_kinetics(low_rate,high_rate,falloff) 
+
+    def to_cantera_kinetics(self, low, high, falloff):
```
Comment (rwest) [inline rmgpy/kinetics/falloff.pyx:421]: This method `to_cantera_kinetics` dosn't refer to `self` and can't actually turn itself into cantera kinetics. i.e. a fully set up `Troe` object `my_troe` cannot do `my_troe.to_cantera_kinetics()` (unlike similar methods in other classes).
I think it should either be deleted and the one line it does just put in the only place it is called (line 419), or else it should be made to function just from `to_cantera_kinetics(self)`.

File: rmgpy/kinetics/arrhenius.pyx [hunk 4]
```diff
@@ -227,10 +227,13 @@ cdef class Arrhenius(KineticsModel):
         """
         self._A.value_si *= factor
 
-    def to_cantera_kinetics(self):
+    def to_cantera_kinetics(self, ArrObj=False):
```
Comment (rwest) [inline rmgpy/kinetics/arrhenius.pyx:230]: not a fan of the name `ArrObj`. Too abbreviated, not used elsewhere, and not in "house style". Maybe `cantera_Arrhenius` or `arrhenius_class`? 

Also the docstrings need further work. Currently says "converts to  cantera Arrhenius object", and "Arrhenius(A,b,E)"  instead of referring to "ArrheniusRate"

File: rmgpy/tools/data/various_kinetics/chem_annotated.inp [hunk 5]
```diff
@@ -172,21 +172,6 @@ HO2(5)+CH3(13)=O2(6)+CH4(15)                        1.000e+00 0.000     0.000
     PLOG/ 100.000   1.500e+11 0.000     1.987    /
     PLOG/ 650.000   2.300e+11 0.200     4.968    /
     PLOG/ 2000.000  3.700e+07 1.340     2.186    /
-DUPLICATE
```
Comment (rwest) [inline rmgpy/tools/data/various_kinetics/chem_annotated.inp:175]: Can you elaborate on why this is needed? I worry it's indicative of a bigger problem, as I fear we have a couple of seed mechanisms that use this horrible sum-of-plog format. (mabye to be addressed in a different PR to this, but I don't want it to be forgotten)

**Review by rwest — COMMENTED** [review]
Thanks Chris. Lots of work here.
 I agree with @JacksonBurns that forcing change to 2.6 and yaml from 2.3 and cti is likely a major change for some users, but I can't see a way for both options to co-exist. (or at least not a way that is simple enough to be worth pursuing). 

Note regarding my review:  I just read the code on github and haven't actually tried running any of it.

File: rmgpy/tools/data/various_kinetics/chem_annotated.inp (hunk 5 above)
Comment (ChrisBNEU) [reply rmgpy/tools/data/various_kinetics/chem_annotated.inp:175]: I will preface this by saying I think the last time the canteramodel tool in rmg was updated was when connie did the uncertainty analysis paper.  So it is outdated, I think before cantera would just ignore bad reactions. 
This is a little complicated. The error in 2.6 is: 
```
CanteraError: 
*******************************************************************************
InputFileError thrown by Kinetics::checkDuplicates:
Error on line 210 of [/Users/blais.ch/Documents/_01_code/RMG_env_1/testing_folder/2022_05_09_cantera_v2_5/chemkin_error.yaml](https://file+.vscode-resource.vscode-cdn.net/Users/blais.ch/Documents/_01_code/RMG_env_1/testing_folder/2022_05_09_cantera_v2_5/chemkin_error.yaml):
No duplicate found for declared duplicate reaction number 7 (CH3(13) + HO2(5) <=> CH4(15) + O2(6))
|  Line |
|   205 |   rate-constants:
|   206 |   - {P: 0.1 atm, A: 8.8e+16, b: -1.05, Ea: 6.461}
|   207 |   - {P: 1.0 atm, A: 8.0e+21, b: -2.39, Ea: 11.18}
|   208 |   - {P: 10.0 atm, A: 3.3e+24, b: -3.04, Ea: 15.61}
|   209 |   note: 'Reaction index: Chemkin #7; RMG #4'
>   210 > - equation: HO2(5) + CH3(13) <=> O2(6) + CH4(15)  # Reaction 8
            ^
|   211 |   duplicate: true
|   212 |   type: pressure-dependent-Arrhenius
|   213 |   rate-constants:
*******************************************************************************
```
I believe this is a bug that is addressed in later versions, but since it is for v3.0, which hasn't been released, I can't find release notes about it. Version 3.0 it accepts the duplicate, but gives the following error: 

```
Invalid rate coefficient for reaction 'CH3(13) + HO2(5) <=> CH4(15) + O2(6)'
at P = 2.0265e+08, T = 200.0
at P = 2.0265e+08, T = 500.0
at P = 2.0265e+08, T = 1000.0
at P = 2.0265e+08, T = 2000.0
at P = 2.0265e+08, T = 5000.0
at P = 2.0265e+08, T = 10000.0
```
This, I think, is expected behavior, as shown in the [plog.cpp program](https://github.com/Cantera/cantera/blob/e82f932119705596410719718e31a4ee63b2e48e/src/kinetics/PlogRate.cpp#L128). It is indicative of a negative rate. the A-factor at 2000 atm (2.0265e+08, Pa) is 0 so that's probably why. if 

both 3.0 and 2.6 function properly if I remove the rate at 2000 atm for reaction 9. further, version 2.6 gives me an "unmarked duplicate reaction" error if I fail to mark both reactions as duplicates, so I think the first error is due to cantera ignoring reaction 9 since it's invalid, and throwing an error for reaction 8, since it no longer has a duplicate. 

What I will do is remove this part of Reaction 9: 
```
PLOG/ 2000.000  0.000e+00 0.000     0.000    /
```
If I remove that both 2.6 and 3.0 can parse the mechanism and reload a `ct.Solution` object no problem. Other people are free to create an issue from  this, I am unsure if rmg has a validation check for library or generated plog reaction rates being negative at certain pressure/temperature combos.

File: rmgpy/kinetics/arrhenius.pyx (hunk 4 above)
Comment (ChrisBNEU) [reply rmgpy/kinetics/arrhenius.pyx:230]: changing to `arrhenius_class`.

Comment (ChrisBNEU) [reply rmgpy/kinetics/arrhenius.pyx:230]: how about this for a docstring on "to_cantera_kinetics"

```
"""
        Converts the RMG Arrhenius object to a cantera ArrheniusRate or
        the auxiliary cantera Arrhenius (used by falloff reactions). 
        Inputs for both are (A,b,E)  where A is in units of m^3/kmol/s, b is dimensionless, and E is in J/kmol

        arrhenius_class: If true, uses cantera.Arrhenius (for falloff reactions). If false, uses 
        Cantera.ArrheniusRate
"""
```

**Review by JacksonBurns — APPROVED** [review]
LGTM. Thanks for making this important PR happen.
